packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,534 @@
+   * @private
+   */
+  async connect() {
+    logger.info('connecting…')
+    await this.retryManager.waitBeforeNextAttempt()
+    this.createSocket()
   }
 
-  if (id && eventName === 'created') {
-    msg = `The 'id' should not be specified for 'created' event.`
+  /**
+   * Throw the previous socket and connect a new one
+   * @private
+   */
+  async reconnect() {
+    if (this.hasWebSocket()) this.rejectWebSocket()
+    this.connect()
   }
 
-  if (typeof handler !== 'function') {
-    msg = `The handler '${handler}' should be a function.`
+  /**
+   * Remove all handlers on a websocket to avoid callbacks from an old rejected socket
+   * @private
+   */
+  rejectWebSocket() {
+    this.emit('disconnected')
+    if (this.hasWebSocket()) {
+      logger.info('trashing the previous websocket…')
+      this.websocket.onmessage = null
+      this.websocket.onerror = null
+      this.websocket.onopen = null
+      this.websocket.onclose = null
+      try {
+        this.websocket.close()
+      } catch (e) {
+        // void
+      }
+      this.websocket = null
+    } else {
+      logger.warn(
+        'Trying to reject a websocket that is not attached. This should not happen'
+      )
+    }
   }
 
-  if (msg) {
-    logger.error(msg)
-    throw new Error(msg)
+  /**
+   * Close the Realtime
+   * @private
+   */
+  stop() {
+    if (this.isStarted) {
+      logger.info('stopped')
+      this.unsubscribeGlobalEvents()
+      this.unsubscribeClientEvents()
+      if (this.hasWebSocket()) {
+        this.rejectWebSocket()
+      }
+      this.isStarted = false
+      if (!this.subscriptions.isEmpty()) {
+        this.unsubscribeAll()
+      }
+      this.emit('close')
+    }
   }
-}
 
-/**
- * Return websocket url from cozyClient
- *
- * @return {String}  WebSocket url
- */
-export const getWebSocketUrl = cozyClient => {
-  const isSecureURL = url => !!url.match(`^(https:/{2})`)
-
-  const url = cozyClient.stackClient.uri
-  const protocol = isSecureURL(url) ? 'wss:' : 'ws:'
-  const host = new URL(url).host
-
-  return `${protocol}//${host}/realtime/`
-}
-
-/**
- * Return token from cozyClient
- *
- * @return {String}  token
- */
-export const getWebSocketToken = cozyClient =>
-  cozyClient.stackClient.token.accessToken || cozyClient.stackClient.token.token
-
-/**
- * CozyRealtime class
- *
- * @class
- */
-class CozyRealtime {
   /**
-   * A cozy client
-   *
-   * @type {CozyClient}
+   * Authenticate to the websocket
+   * @private
    */
-  _cozyClient = null
+  authenticate() {
+    if (this.isWebSocketOpen()) {
+      const token = getToken(this.client)
+      logger.debug('authenticate with', token)
+      this.sendWebSocketMessage('AUTH', token)
+      this.websocket.authenticated = true
+    } else {
+      logger.warn(
+        'Trying to authenticate to a non-opened websocket. This should not happen.',
+        this.websocket,
+        this.websocket && this.websocket.readyState
+      )
```

Comment:
Well, warning should not happen either. The semantic in use is "If we can recover automatically or if we end up in a stable situation which is the one the developper wanted anyways, it's a warning, otherwise it's an error"

Reply:
So it means that in our future work to connect minilog to sentry, we will listen to not only error but also warn? just to keep in mind

Reply:
That's a good question. I didn't have that in mind. @ptbrowne do you have advices on this?

Reply:
I looked on the internet for a mental model of what should be an error and what should be a warning:

> A warning tends to be a notice (either from the compiler or a library) that does not imply a fatal error, but instead something that could go wrong (eg memory leakage, unused variables) or something that might not work in the future (eg deprecation). An error means that something went wrong, either recoverably or unrecoverably. A recoverable error is one after which the program can (probably) still continue whereas an unrecoverable error is one which causes the program to 'hang' or 'crash'.

https://www.quora.com/What-is-the-difference-between-warning-and-errors

I like this mental model. With this mental model, I think here we have an error, even if this is recoverable. As a developer, I look at the error and think that I should fix it, which seems to be the case here.

> So it means that in our future work to connect minilog to sentry, we will listen to not only error but also warn?

I think warnings (notices that something could break) are informative and can help when debugging, so I am OK to send them to Sentry.

Reply:
Thanks for the research. Let's put errors is these cases then.